bolts/fair/models/fair_mixup.py

```diff
@@ -0,0 +1,286 @@
+        return self._inference_step(batch=batch, stage="validate")
+
+    @implements(nn.Module)
+    def forward(self, x: Tensor) -> Tensor:
+        return self.net(x)
+
+    @implements(pl.LightningModule)
+    def test_epoch_end(self, output_results: list[Mapping[str, Tensor]]) -> None:
+        results_dict = self._inference_epoch_end(results=output_results, stage="test")
+        self.log_dict(results_dict)
+
+    @implements(pl.LightningModule)
+    def test_step(self, batch: DataBatch, batch_idx: int) -> dict[str, Tensor]:
+        return self._inference_step(batch=batch, stage="test")
+
+    @implements(pl.LightningModule)
+    def configure_optimizers(
+        self,
+    ) -> tuple[list[optim.Optimizer], list[Mapping[str, LRScheduler | int | SchedInterval]]]:
+        opt = optim.AdamW(
+            self.parameters(),
+            lr=self.learning_rate,
+            weight_decay=self.weight_decay,
+        )
+        sched = {
+            "scheduler": CosineAnnealingWarmRestarts(
+                optimizer=opt, T_0=self.lr_initial_restart, T_mult=self.lr_restart_mult
+            ),
+            "interval": self.lr_sched_interval,
+            "frequency": self.lr_sched_freq,
+        }
+        return [opt], [sched]
+
+    @staticmethod
+    def mixup_data(
+        *,
+        batch: DataBatch,
+        device: torch.device,
+        mix_lambda: Optional[float],
+        alpha: float,
+    ) -> tuple[Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, dict[str, float]]:
+        '''Returns mixed inputs, pairs of targets, and lambda'''
+        lam = (
+            Beta(
+                torch.tensor([alpha]).to(device),
+                torch.tensor([alpha]).to(device)
+                # Potentially change alpha from a=1.0 to account for class imbalance?
+            ).sample()
+            if mix_lambda is None
+            else torch.tensor([mix_lambda]).to(device)
+        )
+
+        batch_s0 = batch.x[batch.s.view(-1) == 0].to(device)
+        batch_s1 = batch.x[batch.s.view(-1) == 1].to(device)
+
+        batch_y_s0 = batch.y[batch.s.view(-1) == 0].to(device)
+        batch_y_s1 = batch.y[batch.s.view(-1) == 1].to(device)
+
+        torch.randperm(batch_s0.size()[0]).to(device)
+        torch.randperm(batch_s1.size()[0]).to(device)
+
+        xal = []
+        xbl = []
+        sal = []
+        sbl = []
+        yal = []
+        ybl = []
+
+        for x_a, s_a, y_a in zip(batch.x, batch.s, batch.y):
+            if s_a == 0:
+                xal.append(x_a)
+                sal.append(s_a.unsqueeze(-1).float())
+                yal.append(y_a.unsqueeze(-1).float())
+                idx = torch.randint(batch_s1.shape[0], (1,))
+                x_b = batch_s1[idx, :].squeeze(0)
+                xbl.append(x_b)
+                sbl.append(torch.ones_like(s_a).unsqueeze(-1).float())
+                y_b = batch_y_s1[idx].float()
+                ybl.append(y_b)
+
+            elif s_a == 1:
```

Comment:
This is the same as the s == 0 branch except for one variable - if you're not going to vectorize the code, you should at least write a local func

Reply:
how would I go about vectorising it? I need a random pairing